core/string/src/builder.rs

```diff
@@ -0,0 +1,442 @@
+use crate::{
+    alloc_overflow, tagged::Tagged, JsString, RawJsString, RefCount, TaggedLen, DATA_OFFSET,
+};
+use std::{
+    alloc::{alloc, dealloc, realloc, Layout},
+    cell::Cell,
+    marker::PhantomData,
+    mem::ManuallyDrop,
+    ops::AddAssign,
+    ptr::{self, addr_of, addr_of_mut, NonNull},
+};
+
+#[doc(hidden)]
+mod private {
+    pub trait Sealed {}
+
+    impl Sealed for u8 {}
+    impl Sealed for u16 {}
+}
+
+/// Inner elements represented for `JsStringBuilder`.
+pub trait JsStringData: private::Sealed {}
+
+impl JsStringData for u8 {}
+impl JsStringData for u16 {}
```

Comment:
I'm actually a bit worried about making the user choose the correct type of string builder instead of just handling that for them.

Can we just assume they want latin1 strings, and automatically switch to UTF-16 if they push either a character outside of the range of latin1 or an u16 value?

Reply:
> Can we just assume they want latin1 strings, and automatically switch to UTF-16 if they push either a character outside of the range of latin1 or an u16 value?

In `build` method there is a check for `latin1`, and with type inference I think most of the time users needn't to manually choose the specific type.

Reply:
Yeah, but this means anyone who wants to use the API must be aware that they need to convert their `str`s into `utf16` format before pushing them. This also means that you cannot mix encodings easily e.g. push latin1, then push an u16 slice, then push more latin1. Ideally the user would be able to just push any `str` into the builder without worrying about encoding UTF-16 or checking for latin1-ness.

Reply:
Then I think this should not put into normal `push`-like methods because `JsStringBuilder` was designed to be used as a faster, and more natural alternative of `Vec`. 
Maybe we can do something like `type MixedEncodingStringBuilder = JsStringBuilder<u16>` and provide different set of API.

Reply:
We don't need to do that, we just need to add a new constructor for `JsStringBuilder` that assumes the UTF-16 encoding. Then, the user will have two options:
- Use `JsStringBuilder::new` without worries, but with the downside of possibly losing some performance if the input has non-latin1 bytes. AKA, easy to use, hard to misuse. They can also use this if they're really sure their input is just ascii or just latin1 characters.
` Use `JsStringBuilder::new_utf16` to squeeze some additional performance if they know the input will have characters outside latin1.

Reply:
If we change that way, those things need to be done:
* Replace generic with a flag to determine which base to allocate.
* Add some conditional checkes for UTF-16 adding methods to change flag.
* Copy the `latin1` part when encoding is changed to `utf16`.

And those things will affect the performance, if we want to keep most of performance and code clean we could do something that restricts users' usage:
* Expose types like `Latin1StringBuilder` and `Utf16StringBuilder` with some extra `string` adding methods instead of exposing `JsStringBuilder`.
* Add some warnings to prevent something like `Latin1StringBuilder.extend_from_slice("Grüße, Jürgen ❤".as_bytes())`.

Reply:
That sounds great! However, we still need a `JsStringBuilder` for people who only want to push `str`s without knowing if the input is Latin1 or UTF-16, and at that point it would be better to just expose some methods like `unchecked_push_latin1/utf16` on `JsStringBuilder`, then make `Latin1/Utf16StringBuilder` wrappers around that which call `unchecked_push_latin1/utf16`

Reply:
I think a common builder like [kiesel-js's work](https://codeberg.org/kiesel-js/kiesel/src/branch/main/src/types/language/String/Builder.zig) would be better, we could export one common builder and other specified byte builders.